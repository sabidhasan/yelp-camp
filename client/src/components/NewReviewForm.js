import React from 'react'
import RatingBar from './RatingBar'

class NewReviewForm extends React.Component {
  constructor(props) {
    super(props);
    //editable is whether or not the form is shown
    //editFormToggle is a parent function that toggles the editable
    this.state = {
      editable: props.editable,
      campgroundID: props.campgroundID,
      // toggleReviewForm: props.toggleReviewForm,
      // addNewComment: props.addNewComment,
      pickedRating: 0,
      reviewText: '',
      errorMessage: null
    }
    // This function sets picked rating for new review
    this.updatePickedRating = this.updatePickedRating.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.postReview = this.postReview.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    // When state is updated in parent
    this.setState({editable: nextProps.editable});
  }

  postReview(event) {

    // authUser.getIdToken()
    // .then(token => {
    //   fetch('/verifyUser', {
    //     method: 'post',
    //     body: JSON.stringify({'userID': token}), //JSON.stringify(token.json),
    //     headers: {
    //       'Accept': 'application/json, text/plain, */*',
    //       'Content-Type': 'application/json'
    //     }
    //   })
    //   .then(response => response.json())
    //   .then(result => console.log(result))
    // })



    event.preventDefault();
    //TO--DO: check for logged in user
    if (!this.state.reviewText) {
      this.setState({errorMessage: 'No review text entered!'});
    } else if (this.state.reviewText.length < 10) {
      this.setState({errorMessage: 'Review too short'})
    } else if (this.state.pickedRating === 0) {
      this.setState({errorMessage: 'Please pick a rating'})
    } else {
      //post to server
      fetch('/comment', {
        method: 'POST',
        headers: {'Accept': 'application/json', 'Content-Type': 'application/json'},
        body: JSON.stringify({
          id: this.state.campgroundID,
          author: 'Jane Doe',
          pickedRating: this.state.pickedRating,
          reviewText: this.state.reviewText})
      })
      .then(res => {
        console.log('res');
        if (!res.ok) throw Error(res.statusText);
        return res.json();
      })
      .then(data => {
        this.props.addNewComment(data);
        //hide the form
        this.setState({pickedRating: 0, reviewText: '', errorMessage: null})
        this.props.toggleReviewForm(null, false);
      })
      .catch(err => {
        this.setState({errorMessage: 'Could not post review at this time. Try again'});
        console.log(err);
      })
    }
  }

  handleChange(event) {
    // change handle for text area
    this.setState({reviewText: event.target.value, errorMessage: null})
  }

  updatePickedRating(newRating) {
    this.setState({pickedRating: newRating})
  }

  render() {
    if (!this.state.editable) return null;
    return (
      <form className='review-form'>
        <span className='review-form__picker rating'>
          <RatingBar rating={this.state.pickedRating} updateRating={this.updatePickedRating} />
        </span>
        <textarea
          required
          className='review-form__text'
          value={this.state.reviewText}
          onChange={this.handleChange}
          placeholder='Add some review stuff here!'
          name='reviewText'
        />
        <span className='error'>{this.state.errorMessage}</span>
        <div className='review-form__buttons'>
          <button className='review-link' onClick={this.postReview}>Post Your Review</button>
          <button onClick={(event) => {
            this.setState({pickedRating: 0, reviewText: '', errorMessage: null});
            this.props.toggleReviewForm(event, false)
          }}>Cancel</button>
        </div>
      </form>
    )
  }
}

export default NewReviewForm;
