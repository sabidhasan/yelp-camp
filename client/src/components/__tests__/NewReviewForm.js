import React from 'react'
import toJson from 'enzyme-to-json'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import NewReviewForm from '../NewReviewForm'

configure({ adapter: new Adapter() });

describe('<NewReviewForm />', () => {
  let component;
  let props;

  let addNewComment;
  let toggleReviewForm;
  let idTokenSuccess;
  const event = { preventDefault() {} }

  const getIdToken = () => {
    return new Promise((res, rej) => {
      if (!idTokenSuccess) {
        return rej()
      }
      return res();
    })
  }

  beforeEach(() => {
    idTokenSuccess = true;
    addNewComment = jest.fn();
    toggleReviewForm = jest.fn();

    props = {
      campgroundID: 11,
      addNewComment,
      toggleReviewForm,
    }
    component = shallow(<NewReviewForm { ...props } />, {
      context:{
        user: {
          photoURL: 'url', displayName: 'Name', loading: false, getIdToken
        }
      }
    })
  })

  it('Component should be defined', () => {
    expect(component).toBeDefined();
  });

  it('Author name is displayed', () => {
    expect(component.find('.NewReviewForm__author').text()).toBe('Posting publically as Name')
  })

  it('Textarea onChange changes value', () => {
    component.find('.NewReviewForm__text').simulate('change', {target: {value: 'New text'}})
    expect(component.find('.NewReviewForm__text').props().value).toBe('New text')
  })

  it('When clicking Post button, if there is no review text, error shown', () => {
    component.find('.NewReviewForm__post-review').simulate('click', event)
    expect(component.find('.NewReviewForm__error').text()).toBe('No review text was entered!')
  })

  it('When clicking Post button, if review text is too short, error is shown', () => {
    component.setState({reviewText: 'abcd'})
    component.find('.NewReviewForm__post-review').simulate('click', event)
    expect(component.find('.NewReviewForm__error').text()).toMatch(/Review is too short/)
  })

  it('When clicking Post button, if no rating is picked, error is shown', () => {
    component.setState({reviewText: '1234567890123456'})
    component.find('.NewReviewForm__post-review').simulate('click', event)
    expect(component.find('.NewReviewForm__error').text()).toBe('Please pick a rating')
  })

  it('When clicking Post button, if user isn\'t logged in, error is shown', () => {
    component.setContext({ user: undefined })
    component.find('.NewReviewForm__post-review').simulate('click', event)
    expect(component.find('.NewReviewForm__error').text()).toMatch(/Sorry, could not post/)
  })

  it('If you create an error, then type something in textbox, error is reset', () => {
    component.setState({errorMessage: 'some error'});
    component.find('.NewReviewForm__text').simulate('change', {target: {value: 'New text'}})
    expect(component.find('.NewReviewForm__error').text()).toBe('')
  })

  it('If you create an error, then pick a new rating, error is reset', () => {
    component.setState({errorMessage: 'some error'});
    component.instance().updatePickedRating(4);
    expect(component.find('.NewReviewForm__error').text()).toBe('')
  })

  it('Clicking cancel calls toggleReviewForm', () => {
    component.find('.NewReviewForm__cancel').simulate('click');
    expect(toggleReviewForm).toBeCalledWith(undefined, false)
  })

  it('Clicking cancel resets form text', () => {
    component.setState({reviewText: 'some text here'})
    component.find('.NewReviewForm__cancel').simulate('click');
    expect(component.find('.NewReviewForm__text').props().value).toBe('')
  })

  it('Clicking cancel resets RatingBar rating to 0', () => {
    component.setState({pickedRating: 4})
    component.find('.NewReviewForm__cancel').simulate('click');
    expect(component.find('RatingBar').props().rating).toBe(0)
  })

  it('Clicking cancel resets error text', () => {
    component.setState({errorMessage: 'there is an error'})
    component.find('.NewReviewForm__cancel').simulate('click');
    expect(component.find('.NewReviewForm__error').text()).toBe('')
  })

  it('If review text is too long, proper warning text is shown', () => {
    const longString = 'qwertyuiopasdfghjklzxcvbnm1234567890.-=[];,~!,~!@#'.repeat(19);
    component.setState({reviewText: longString})
    expect(component.find('.NewReviewForm__warning').text()).toBe('50 characters left');
  })

  it('NewReviewForm component snapshot matches what is expected', () => {
    expect(toJson(component)).toMatchSnapshot();
  })
})
