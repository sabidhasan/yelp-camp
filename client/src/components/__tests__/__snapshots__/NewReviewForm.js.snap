// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewReviewForm /> NewReviewForm component snapshot matches what is expected 1`] = `
<form
  className="NewReviewForm"
  role="textbox"
>
  <RatingBar
    rating={0}
    updateRating={[Function]}
  />
  <span
    className="NewReviewForm__author bold"
  >
    Posting publically as 
    Name
    <img
      alt="User icon"
      className="NewReviewForm__image"
      src="url"
    />
  </span>
  <textarea
    aria-invalid={false}
    aria-required="true"
    className="NewReviewForm__text"
    maxLength={1000}
    name="reviewText"
    onChange={[Function]}
    placeholder="Write your honest review here!"
    required={true}
    value=""
  />
  <span
    aria-relevant="all"
    className="NewReviewForm__error bold"
    role="alert"
  />
  <div
    className="NewReviewForm__controls"
  >
    <button
      aria-disabled="false"
      className="NewReviewForm__post-review btn btn--small"
      disabled={false}
      onClick={[Function]}
    >
      Post Your Review
    </button>
    <button
      className="NewReviewForm__cancel btn btn--small"
      onClick={[Function]}
    >
      Cancel
    </button>
    <div
      className="NewReviewForm__progress flex-center bold"
    >
      <t
        background={false}
        backgroundPadding={null}
        classForPercentage={null}
        className=""
        classes={
          Object {
            "background": "CircularProgressbar-background",
            "path": "CircularProgressbar-path",
            "root": "CircularProgressbar",
            "text": "CircularProgressbar-text",
            "trail": "CircularProgressbar-trail",
          }
        }
        counterClockwise={false}
        initialAnimation={false}
        percentage={0}
        strokeWidth={15}
        styles={
          Object {
            "path": Object {
              "stroke": null,
            },
          }
        }
        textForPercentage={[Function]}
      />
    </div>
  </div>
</form>
`;
